crates/ibc/src/services/core/context.rs

```diff
@@ -0,0 +1,93 @@
+use crate::core::ics02_client::client_state::Status;
+use crate::core::ics02_client::client_type::ClientType;
+use crate::core::ics03_connection::connection::IdentifiedConnectionEnd;
+use crate::core::ics04_channel::channel::IdentifiedChannelEnd;
+use crate::core::ics04_channel::packet::Sequence;
+use crate::core::ics24_host::identifier::{ClientId, ConnectionId};
+use crate::core::ics24_host::path::{AckPath, ChannelEndPath, CommitmentPath, Path};
+use crate::core::{ContextError, ValidationContext};
+use crate::prelude::*;
+use crate::Height;
+
+/// Context to be implemented by the host to provide proofs in gRPC query responses
+///
+/// Trait used for the [`gRPC query services`](crate::services).
+pub trait ProvableContext {
+    /// Returns the proof for the given path at the given height.
+    /// As this is in the context of IBC, the path is expected to be an [`IbcPath`](Path).
+    fn get_proof(&self, height: Height, path: &Path) -> Option<Vec<u8>>;
+}
+
+/// Context to be implemented by the host that provides gRPC query services.
+///
+/// Trait used for the [`gRPC query services`](crate::services).
+pub trait QueryContext: ProvableContext + ValidationContext {
+    // Client queries
+
+    /// Returns the list of all clients.
+    fn client_states(
+        &self,
+    ) -> Result<Vec<(ClientId, <Self as ValidationContext>::AnyClientState)>, ContextError>;
+
+    /// Returns the list of all consensus states for the given client.
+    fn consensus_states(
+        &self,
+        client_id: &ClientId,
+    ) -> Result<Vec<(Height, <Self as ValidationContext>::AnyConsensusState)>, ContextError>;
+
+    /// Returns the list of all heights at which consensus states for the given client are.
+    fn consensus_state_heights(&self, client_id: &ClientId) -> Result<Vec<Height>, ContextError>;
+
+    /// Returns the status of the given client.
+    fn client_status(&self, client_id: &ClientId) -> Result<Status, ContextError>;
```

Comment:
I see we can access the client status as follows, so this method can be dropped.
```rust
let client_state = self.ibc_context.client_state(&client_id)?;
let client_validation_ctx = self.ibc_context.get_client_validation_context();
let client_status = client_state
  .status(&client_validation_ctx, &client_id)
  .map_err(...)?;
```

Reply:
very good catch!